### ISSUE TITLE
[Feature] Tree Table Widget

### ISSUE THREAD

**Issue by Nikhil-Nandagopal** [appsmithorg/appsmith#5415]
## Summary

A user would like a tree view of the table as shown where each row can be collapsed / expanded.
https://ant.design/components/table/#header

**Comment by somangshu** [comment]
Should this be considered in new widget or a functionality of the current table widget?

**Comment by Nikhil-Nandagopal** [comment]
@somangshu 
does the underlying component support it as a prop or a separate component?
Is it easy to configure in the property pane?
My guess is a separate widget is better but as always let's fall back to the above 2 questions ☝️

**Comment by somangshu** [comment]
@Nikhil-Nandagopal while other design frameworks will tell us that the feature can be under the same component, I feel that the UX in the Appsmith environment will be very complicated (answer to the second question). 

In favour of the above, I would also consider this to be a separate widget!

**Comment by areyabhishek** [comment]
How would the user choose what to show when the row is expanded? This feels like a table control, but for discovery we can treat it as a new widget.

**Comment by Nikhil-Nandagopal** [comment]
@somangshu it need not be very complicated. We just need a property to enable horizontal scroll / collapse columns. If horizontal scroll is enabled, all columns that cannot be fit on the screen can be scrolled to be viewed. If Collapse columns is enabled, all columns that do not fit the table width are collapsed into the + option

**Comment by Nikhil-Nandagopal** [comment]
@somangshu yes horizontal scroll is present. It is used when the user wants to see additional columns that cannot be shown in the table. Expand row also does the same thing, it is used to show additional columns of rows that cannot be shown in the table. Thus the two are simply 2 modes of achieving the same goal via different UX. One where you scroll to see the next column and one where you expand the row to see the next column.

**Comment by somangshu** [comment]
Got it. I was thinking about showing additional information against each row. This should be fairly simple them

**Comment by somangshu** [comment]
note:
- New prop: collapse columns - toggle; Table wide setting
    - Enabled: No horizontal scroll and the rest of the data appears as an accordion when the row data is expanded 
    - Disabled: all additional columns can be viewed by scrolling horizontally 
    -

**Comment by somangshu** [comment]
@Tooluloope we need to triage if its possible to do this in the current table widget implementation

**Comment by Tooluloope** [comment]
From my findings, the Current Table Library doesn't support showing columns as it can only support showing nested rows and not column.

**Comment by Jonsey4** [comment]
I have a single sql table of items which may optionally have a parent and/or child in the same table. This kind of widget would be great for visually modelling these parent-child relationships without hacking something together with a single table widget.

**Comment by pchambless** [comment]
Will there be some requirement around how to represent a hierarchy in this widget?  Like, the **Select** widget needs columns named **value** and **label.**.  Some hierarchies use Integer (non-intelligent) ids while some people like to use intelligent IDs to represent their hierarchy parent-child relationships... I guess this could be some sort of property dropdown... like when you define chart **Series**?

**Comment by somangshu** [comment]
Thanks for the context @Jonsey4 @pchambless 

cc @dilippitchika FYI, something to consider

**Comment by ame-appsmith** [comment]
[Another user on Discord](https://discord.com/channels/725602949748752515/1022133601900445786) requested this feature.

**Comment by brusamatteo** [comment]
Really interested in this feature. The ability to show a BOM tree is very usefull

**Comment by zuroma999** [comment]
This would be a great feature. Other low-coder systems I'm evaluating allow you to drag and drop widgets into the expanded area beneath each row, to build out whatever needed, and the widgets access the same context (table.currentRow) etc.

<img width="1820" alt="Screenshot 2024-01-17 at 10 13 02 AM" src="https://github.com/appsmithorg/appsmith/assets/49953523/36e5100f-2cd6-4272-b1b9-e6eca157acbf">

**Comment by Nikhil-Nandagopal** [comment]
@zuroma999 thanks for dropping that comment. I'd be interested to learn about what you're building and how you plan on using this widget. Would you be up for a quick chat? You can find a convenient time on my calendar below

https://calendly.com/appsmith-nikhil/30min?back=1&month=2024-01

**Comment by zuroma999** [comment]
@Nikhil-Nandagopal, sure, I'll book something on your calendar to discuss.

**Comment by amirmf** [comment]
any updates?

**Comment by rahulbarwal** [comment]
Hey @amirmf, This is not a priority for the team at the moment. We've logged your interest and would like to understand your usecase, can you share more about it?

We will share when we prioritize this.
